// tls-client/dist/esm/index.js
import { Session } from './sessions';
import { Response } from './response';
import { TLSClientException } from './exceptions';
const TlsClient = {
    Session,
    Response,
    TLSClientException,
    async get(url, options) {
        const session = new Session();
        return await session.get(url, options);
    },
    async post(url, options) {
        const session = new Session();
        return await session.post(url, options);
    },
    async put(url, options) {
        const session = new Session();
        return await session.put(url, options);
    },
    async patch(url, options) {
        const session = new Session();
        return await session.patch(url, options);
    },
    async delete(url, options) {
        const session = new Session();
        return await session.delete(url, options);
    },
    async head(url, options) {
        const session = new Session();
        return await session.head(url, options);
    },
    async options(url, options) {
        const session = new Session();
        return await session.options(url, options);
    }
};
export default TlsClient;
module.exports = TlsClient;
